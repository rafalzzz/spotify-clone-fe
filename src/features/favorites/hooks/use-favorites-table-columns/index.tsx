'use client';
import type { ColumnsType } from 'antd/es/table';
import { useMemo } from 'react';

import { ContextMenuCell } from '@/favorites/components/context-menu-cell';

import { TSongItem } from '@/types/components';
import { EMusicTrackKeys } from '@/types/music-track';

export const useFavoritesTableColumns = (): ColumnsType<TSongItem> => {
  const columns = useMemo(
    (): ColumnsType<TSongItem> => [
      {
        title: '#',
        key: EMusicTrackKeys.TRACK_ID,
        render: (_, record, index) => (
          <ContextMenuCell record={record}>
            <div style={{ width: '100%', height: '100%' }}>{index}</div>
          </ContextMenuCell>
        ),
      },
      /*   {
      title: 'Title',
      key: EMusicTrackKeys.TRACK_ID,
      render: (_, record) => (
        <ContextMenuCell record={record}>
          <div style={{ width: '100%', height: '100%' }}>{record[EMusicTrackKeys.TRACK_NAME]}</div>
        </ContextMenuCell>
      ),
    }, */
    ],
    [],
  );

  return columns;
};
