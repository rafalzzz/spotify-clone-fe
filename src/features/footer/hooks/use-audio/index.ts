import { useRef, useCallback, useEffect, useMemo, SyntheticEvent } from 'react';

import { useMusicPlayerContext } from '@/footer/contexts/music-player-context';
import { TUseAudio, TUseAudioProps } from '@/footer/types';

import { useMusicPlayerStore } from '@/store/music-player';

export const useAudio = ({ setCurrentTime }: TUseAudio): TUseAudioProps => {
  const { ref, isLoop } = useMusicPlayerContext();
  const { isPlaying, albumId, activeIndex, songsList, togglePlay, setDuration, setActiveIndex } =
    useMusicPlayerStore();

  const lastUpdatedTime = useRef(0);

  const currentSong = useMemo(() => songsList[activeIndex], [activeIndex, songsList]);

  const onTimeUpdate = useCallback(
    ({ target }: SyntheticEvent<HTMLAudioElement>) => {
      const { currentTime } = target as HTMLAudioElement;

      // The purpose of the following procedure is to reduce the number
      // of re-renders when updating the playback time of the song being played.
      const roundedCurrentTime = Math.floor(currentTime);

      if (roundedCurrentTime !== lastUpdatedTime.current) {
        setCurrentTime(roundedCurrentTime);
        lastUpdatedTime.current = roundedCurrentTime;
      }
    },
    [setCurrentTime],
  );

  const onLoadedMetadata = useCallback(
    ({ target }: SyntheticEvent<HTMLAudioElement>) => {
      const { duration } = target as HTMLAudioElement;
      setDuration(duration);
    },
    [setDuration],
  );

  const handlePlayingAlbum = useCallback(
    (activeIndex: number, isLastSong: boolean) => {
      setCurrentTime(0);
      setActiveIndex(isLastSong ? 0 : activeIndex + 1);
    },
    [setCurrentTime, setActiveIndex],
  );

  const onEnded = useCallback(() => {
    const isLastSong = activeIndex === songsList.length - 1;
    const isPlayingAlbum = !!albumId;

    if (isPlayingAlbum) {
      return handlePlayingAlbum(activeIndex, isLastSong);
    }

    if (!isLoop && isLastSong) {
      togglePlay();
    }
  }, [activeIndex, albumId, isLoop, songsList.length, handlePlayingAlbum, togglePlay]);

  const handlePlaying = useCallback(() => {
    if (isPlaying) {
      ref.current?.play();
    } else {
      ref.current?.pause();
    }
  }, [isPlaying, ref]);

  useEffect(handlePlaying, [handlePlaying]);

  return { ref, currentSong, isPlaying, isLoop, onLoadedMetadata, onTimeUpdate, onEnded };
};
