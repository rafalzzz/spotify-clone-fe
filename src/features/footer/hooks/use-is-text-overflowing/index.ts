'use client';

import { useState, useRef, useEffect } from 'react';

import { TUseIsTextOverflowing, TUseIsTextOverflowingProps } from '@/footer/types';

export const useIsTextOverflowing = ({
  currentSong,
}: TUseIsTextOverflowing): TUseIsTextOverflowingProps => {
  const [isTextOverflowing, setIsTextOverflowing] = useState(false);

  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (ref.current) {
      const titleRect = ref.current.getBoundingClientRect();
      const parentNode = ref.current.parentNode as HTMLDivElement;
      const containerRect = parentNode.getBoundingClientRect();

      const isOverflow = titleRect.width > containerRect.width;
      setIsTextOverflowing(isOverflow);
    }
  }, [currentSong]);

  return { ref, isTextOverflowing };
};
