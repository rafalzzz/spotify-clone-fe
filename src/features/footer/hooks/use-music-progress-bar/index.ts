import { useState, useRef, useMemo, ChangeEvent, useCallback } from 'react';

import { useMusicPlayerContext } from '@/footer/contexts/music-player-context';
import { TUseMusicProgressBarProps } from '@/footer/types';

import { useMusicPlayerStore } from '@/store/music-player';

export const useMusicProgressBar = (): TUseMusicProgressBarProps => {
  const [currentTime, setCurrentTime] = useState(0);
  const [temporaryTime, setTemporaryTime] = useState<number | null>(null);
  const [isReversedTime, setIsReversedTime] = useState(false);

  const { ref } = useMusicPlayerContext();
  const { duration } = useMusicPlayerStore();

  const isChangingRef = useRef(false);

  const durationValue = useMemo(
    () => (isReversedTime ? duration - currentTime : duration),
    [currentTime, duration, isReversedTime],
  );

  const handleStartChange = useCallback(() => {
    isChangingRef.current = true;
  }, []);

  const handleChange = useCallback(
    ({ target: { valueAsNumber } }: ChangeEvent<HTMLInputElement>) => {
      if (isChangingRef.current) {
        setTemporaryTime(valueAsNumber);
      }
    },
    [],
  );

  const handleEndChange = useCallback(() => {
    if (ref.current && temporaryTime) {
      isChangingRef.current = false;
      ref.current.currentTime = temporaryTime;
      setTemporaryTime(null);
    }
  }, [ref, temporaryTime]);

  const onClick = useCallback(() => {
    setIsReversedTime((prevState) => !prevState);
  }, []);

  return {
    duration,
    currentTime,
    temporaryTime,
    durationValue,
    isReversedTime,
    setCurrentTime,
    handleStartChange,
    handleChange,
    handleEndChange,
    onClick,
  };
};
