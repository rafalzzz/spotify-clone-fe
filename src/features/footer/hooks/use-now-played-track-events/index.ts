import { useState } from 'react';

import { TUseMusicTrackTitleTooltipProps } from '@/footer/types';

export const useMusicTrackTitleTooltips = (): TUseMusicTrackTitleTooltipProps => {
  const [isMenuOpen, setIsMenuOpen] = useState(false);
  const [isMouseOver, setIsMouseOver] = useState(false);

  const onOpenChange = (open: boolean) => {
    setIsMenuOpen(open);
  };

  const onMouseEnter = () => {
    setIsMouseOver(true);
  };

  const onMouseLeave = () => {
    setIsMouseOver(false);
  };

  return { isMenuOpen, isMouseOver, onOpenChange, onMouseEnter, onMouseLeave };
};
