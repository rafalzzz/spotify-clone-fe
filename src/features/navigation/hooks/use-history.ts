'use-client';
import { usePathname } from 'next/navigation';
import { useReducer, useCallback, useEffect } from 'react';

enum HistoryActionKind {
  UNDO_BUTTON = 'UNDO_BUTTON',
  REDO_BUTTON = 'REDO_BUTTON',
  ADD_PATH = 'ADD_PATH',
  UPDATE_PATHS = 'UPDATE_PATHS',
  UPDATE_PATH_INDEX = 'UPDATE_PATH_INDEX',
}

type AddPathAction = { currentPath: string };
type UpdatePathIndexAction = { value: number };

type CountAction = {
  type: HistoryActionKind;
  payload?: AddPathAction | UpdatePathIndexAction;
};

type CountState = {
  allPaths: string[];
  currentPathIndex: number;
};

function historyReducer(state: CountState, action: CountAction) {
  const { type, payload } = action;
  switch (type) {
    case HistoryActionKind.UNDO_BUTTON:
      return {
        ...state,
        currentPathIndex: state.currentPathIndex - 1,
      };
    case HistoryActionKind.REDO_BUTTON:
      return {
        ...state,
        currentPathIndex: state.currentPathIndex + 1,
      };
    case HistoryActionKind.ADD_PATH:
      return {
        ...state,
        currentPathIndex: state.currentPathIndex + 1,
        allPaths: [...state.allPaths, (payload as AddPathAction)?.currentPath],
      };
    case HistoryActionKind.UPDATE_PATHS:
      return {
        ...state,
        currentPathIndex: state.currentPathIndex + 1,
        allPaths: [
          ...state.allPaths.slice(0, state.currentPathIndex),
          (payload as AddPathAction)?.currentPath,
        ],
      };
    case HistoryActionKind.UPDATE_PATH_INDEX:
      return {
        ...state,
        currentPathIndex: state.currentPathIndex + (payload as UpdatePathIndexAction)?.value,
      };
    default:
      return state;
  }
}

const INITIAL_STATE = {
  allPaths: [],
  currentPathIndex: -1,
};

const DISABLED_UNDO_BUTTON_INDEX_VALUES = [-1, 0];

export const useHistory = () => {
  const [state, dispatch] = useReducer(historyReducer, INITIAL_STATE);
  const pathname = usePathname();

  const { allPaths, currentPathIndex } = state;

  const isUndoButtonDisabled = DISABLED_UNDO_BUTTON_INDEX_VALUES.includes(currentPathIndex);
  const isRedoButtonDisabled = currentPathIndex === allPaths.length - 1;

  const undoHistory = useCallback(() => {
    dispatch({ type: HistoryActionKind.UNDO_BUTTON });
  }, []);

  const redoHistory = useCallback(() => {
    dispatch({ type: HistoryActionKind.REDO_BUTTON });
  }, []);

  const handlePathname = useCallback(() => {
    if (!pathname) return;

    const isPathInHistory = allPaths.includes(pathname);
    const isLatestPath = currentPathIndex === allPaths.length - 1;
    const isPrevPath = pathname === allPaths[currentPathIndex - 1];

    switch (true) {
      case !isPathInHistory:
        dispatch({ type: HistoryActionKind.ADD_PATH, payload: { currentPath: pathname } });
        break;

      case isLatestPath && !isPrevPath:
        dispatch({ type: HistoryActionKind.ADD_PATH, payload: { currentPath: pathname } });
        break;

      case !isLatestPath && !isPathInHistory:
        dispatch({ type: HistoryActionKind.UPDATE_PATHS, payload: { currentPath: pathname } });
        break;

      default:
        dispatch({
          type: HistoryActionKind.UPDATE_PATH_INDEX,
          payload: { value: isPrevPath ? -1 : 1 },
        });
        break;
    }
    // This is essential for the proper functioning of the hook.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [pathname]);

  useEffect(handlePathname, [handlePathname]);

  return { isUndoButtonDisabled, isRedoButtonDisabled, undoHistory, redoHistory };
};
