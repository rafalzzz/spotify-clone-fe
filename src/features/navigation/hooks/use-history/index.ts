'use client';

import { usePathname, useRouter } from 'next/navigation';
import { useReducer, useCallback, useEffect } from 'react';

import {
  EHistoryActionKind,
  TAddPath,
  TCountAction,
  TCountState,
  TUpdatePathIndex,
} from '@/navigation/types';

const historyReducer = (state: TCountState, action: TCountAction) => {
  const { type, payload } = action;
  switch (type) {
    case EHistoryActionKind.UNDO_BUTTON:
      return {
        ...state,
        currentPathIndex: state.currentPathIndex - 1,
      };
    case EHistoryActionKind.REDO_BUTTON:
      return {
        ...state,
        currentPathIndex: state.currentPathIndex + 1,
      };
    case EHistoryActionKind.ADD_PATH:
      return {
        ...state,
        currentPathIndex: state.currentPathIndex + 1,
        allPaths: [...state.allPaths, (payload as TAddPath)?.pathname],
      };
    case EHistoryActionKind.UPDATE_PATHS:
      return {
        ...state,
        currentPathIndex: state.currentPathIndex + 1,
        allPaths: [
          ...state.allPaths.slice(0, state.currentPathIndex + 1),
          (payload as TAddPath)?.pathname,
        ],
      };
    case EHistoryActionKind.UPDATE_PATH_INDEX:
      return {
        ...state,
        currentPathIndex: state.currentPathIndex + (payload as TUpdatePathIndex)?.value,
      };
    default:
      return state;
  }
};

const DISABLED_UNDO_BUTTON_INDEX_VALUES = [-1, 0];

type TUseHistory = {
  initialState: {
    allPaths: string[];
    currentPathIndex: number;
  };
};

export const useHistory = ({ initialState }: TUseHistory) => {
  const [state, dispatch] = useReducer(historyReducer, initialState);
  const { push } = useRouter();
  const pathname = usePathname();

  const { allPaths, currentPathIndex } = state;

  const isUndoButtonDisabled = DISABLED_UNDO_BUTTON_INDEX_VALUES.includes(currentPathIndex);
  const isRedoButtonDisabled = currentPathIndex === allPaths.length - 1;

  const undoHistory = useCallback(() => {
    dispatch({ type: EHistoryActionKind.UNDO_BUTTON });
  }, []);

  const redoHistory = useCallback(() => {
    dispatch({ type: EHistoryActionKind.REDO_BUTTON });
  }, []);

  const handlePathname = useCallback(() => {
    if (!pathname) return;

    const isPathInHistory = allPaths.includes(pathname);
    const lastIndex = allPaths.length - 1;
    const isLatestPath = currentPathIndex === lastIndex;
    const isPrevPath = pathname === allPaths[currentPathIndex - 1];
    const isNextPath = pathname === allPaths[currentPathIndex + 1];

    switch (true) {
      case (isLatestPath && !isPathInHistory) || (isLatestPath && !isPrevPath):
        dispatch({ type: EHistoryActionKind.ADD_PATH, payload: { pathname } });
        break;

      case !isLatestPath && !isPathInHistory && !isNextPath:
        dispatch({ type: EHistoryActionKind.UPDATE_PATHS, payload: { pathname } });
        break;

      default:
        dispatch({
          type: EHistoryActionKind.UPDATE_PATH_INDEX,
          payload: { value: isPrevPath ? -1 : 1 },
        });
        break;
    }
    // This is essential for the proper functioning of the hook.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [pathname]);

  useEffect(handlePathname, [handlePathname]);

  const handleHistoryChange = useCallback(() => {
    const historyPathname = allPaths[currentPathIndex];
    if (historyPathname && pathname !== historyPathname) {
      push(allPaths[currentPathIndex]);
    }
    // This is essential for the proper functioning of the hook,
    // because changes connected with `pathname` are handled in useEffect above.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [allPaths, currentPathIndex, push]);

  useEffect(handleHistoryChange, [handleHistoryChange]);

  return { isUndoButtonDisabled, isRedoButtonDisabled, undoHistory, redoHistory };
};
