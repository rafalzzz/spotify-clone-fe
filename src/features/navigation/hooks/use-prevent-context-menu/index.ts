import { useCallback, useEffect } from 'react';

export const usePreventContextMenu = (): void => {
  const preventContextMenu = useCallback((event: MouseEvent) => {
    event.preventDefault();
  }, []);

  useEffect(() => {
    document.body.addEventListener('contextmenu', preventContextMenu);

    return () => {
      document.body.removeEventListener('contextmenu', preventContextMenu);
    };
  }, [preventContextMenu]);
};
