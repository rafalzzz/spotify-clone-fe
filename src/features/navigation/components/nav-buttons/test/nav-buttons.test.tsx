import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import '@testing-library/jest-dom/extend-expect';

import { useHistory } from '@/navigation/hooks/use-history';

import { NavButtons } from '../';

const mockUndoHistory = jest.fn();
const mockRedoHistory = jest.fn();

const UNDO_HISTORY_TEST_ID = 'undo-history';
const REDO_HISTORY_TEST_ID = 'redo-history';

const BASIC_HOOK_VALUES = {
  undoHistory: mockUndoHistory,
  redoHistory: mockRedoHistory,
};

jest.mock('@/navigation/hooks/use-history');

const renderNavButtons = () => render(<NavButtons />);

describe('NavButtons', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('render component without error', () => {
    (useHistory as jest.Mock).mockReturnValue(BASIC_HOOK_VALUES);

    const screen = renderNavButtons();
    expect(screen).toMatchSnapshot();
  });

  it('should call mockUndoHistory when user click undo-history button', async () => {
    (useHistory as jest.Mock).mockReturnValue(BASIC_HOOK_VALUES);

    const { queryByTestId } = renderNavButtons();
    const undoHistoryButton = queryByTestId(UNDO_HISTORY_TEST_ID);
    expect(undoHistoryButton).toBeInTheDocument();

    await userEvent.click(undoHistoryButton as Element);

    expect(mockUndoHistory).toHaveBeenCalledTimes(1);
    expect(mockRedoHistory).not.toHaveBeenCalled();
  });

  it('should call mockRedoHistory when user click redo-history button', async () => {
    (useHistory as jest.Mock).mockReturnValue(BASIC_HOOK_VALUES);

    const { queryByTestId } = renderNavButtons();
    const redoHistoryButton = queryByTestId(REDO_HISTORY_TEST_ID);
    expect(redoHistoryButton).toBeInTheDocument();

    await userEvent.click(redoHistoryButton as Element);

    expect(mockRedoHistory).toHaveBeenCalledTimes(1);
    expect(mockUndoHistory).not.toHaveBeenCalled();
  });
});
