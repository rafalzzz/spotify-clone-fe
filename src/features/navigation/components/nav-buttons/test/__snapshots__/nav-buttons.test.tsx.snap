// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavButtons render component without error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="nav-buttons"
      >
        <li>
          <button
            class="ant-btn css-dev-only-do-not-override-12upa3x ant-btn-circle ant-btn-default nav-buttons__item"
            data-testid="undo-history"
            role="button"
            type="button"
          >
            <span
              aria-label="left"
              class="anticon anticon-left"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="left"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                />
              </svg>
            </span>
          </button>
        </li>
        <li>
          <button
            class="ant-btn css-dev-only-do-not-override-12upa3x ant-btn-circle ant-btn-default nav-buttons__item"
            data-testid="redo-history"
            role="button"
            type="button"
          >
            <span
              aria-label="right"
              class="anticon anticon-right"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="right"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                />
              </svg>
            </span>
          </button>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="nav-buttons"
    >
      <li>
        <button
          class="ant-btn css-dev-only-do-not-override-12upa3x ant-btn-circle ant-btn-default nav-buttons__item"
          data-testid="undo-history"
          role="button"
          type="button"
        >
          <span
            aria-label="left"
            class="anticon anticon-left"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="left"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
              />
            </svg>
          </span>
        </button>
      </li>
      <li>
        <button
          class="ant-btn css-dev-only-do-not-override-12upa3x ant-btn-circle ant-btn-default nav-buttons__item"
          data-testid="redo-history"
          role="button"
          type="button"
        >
          <span
            aria-label="right"
            class="anticon anticon-right"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="right"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
              />
            </svg>
          </span>
        </button>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
